Add `include_prerelease` and `loose` option to version range expression

climblinne opened the pull request:
Require node-semver 0.6.1
Usage example: [~1.2.2,include_prerelease=True,loose=False]

Changelog: (Feature): Add `include_prerelease` and `loose` option to version range expression

- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [X] I've read the [Contributing guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [X] I've followed the PEP8 style guides for Python code.
- [x] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. See https://github.com/conan-io/docs/issues/937
- [ ] Wait for #3964
- [x] Add unit tests

Connects to issue #3798

memsharded commented on conans/client/graph/range_resolver.py:
```diff
@@ -6,18 +6,34 @@
+    version_range = versionexpr.split(",")
+
+    include_prerelease = False
+    if "include_prerelease=True" in version_range:
```
I'd say this can be simplified, and assume "prerelease" means enabling them (nothing to keep current behavior)

climblinne replied:
Changed to `prerelease`

memsharded commented on conans/client/graph/range_resolver.py:
```diff
@@ -6,18 +6,34 @@
+        version_range.remove("include_prerelease=False")
+
+    loose = True
+    if "loose=True" in version_range:
```
Same for "loose", I would use its existence to activate.

climblinne replied:
Changed keyword to `strict`

memsharded commented on conans/test/model/version_ranges_test.py:
```diff
@@ -61,9 +77,9 @@ def basic_test(self):
-        result = satisfying(["1.a.1", "master", "X.2", "1.2.1", "1.3", "2.1"], "1.3", output)
-        self.assertIn("Version 'master' is not semver", output)
-        self.assertEqual(result, "1.3")
+#       result = satisfying(["1.a.1", "master", "X.2", "1.2.1", "1.3", "2.1"], "1.3", output)
```
why commented out?

climblinne replied:
Not a testing artifact. Sorry. There seems to be a change or bug in node-semver.

climblinne replied:
I added a patch for Semver (https://github.com/podhmo/python-semver/pull/24), to get the same result as before. Then we can uncomment the line again.

lasote commented on conans/client/graph/range_resolver.py:
```diff
@@ -12,18 +12,16 @@ def satisfying(list_versions, versionexpr, output):
-        version_range.remove("loose=False")
-        loose = False
+
+    for keyword in version_range:
```
I would prefer something like (not tested):
 
```python
tokens = versionexpr.rsplit(",", 1)
if len(tokens) == 2:
    version_range, operation = tokens
elif len(tokens) == 1:
    version_range, operation = versionexpr, None

include_prerelease = operation == "prerelease"
strict = operation == "strict"
```

climblinne replied:
But we can have 
[>=1.0.0, < 2.0.0, strict, prerelease] or [~1.2.0, prerelease]
So 2-4 tokens.
Anyway, how would we write this range: [1.2.7 || >=1.2.9 <2.0.0], as we normally have to write commas in between?

How about using [>=1.0.0,< 2.0.0; strict; prerelease]? So separating operation with semicolons?

lasote replied:
Maybe it is worth to use a separator when the operations start. But I'm not sure, to be honest.

lasote replied:
@conan-io/barbarians feedback needed

climblinne replied:
Any feedback?

climblinne replied:
@lasote: I still wait for some feedback, could you support?

lasote replied:
@conan-io/barbarians @memsharded @danimtb @jgsogo Feedback.

jgsogo replied:
It seems more convenient to have `[<version_range>, <**options>]`, although I think we have to be compatible with existing behavior, which according to the sources is (relevant snippets):

```python
    if version.startswith("[") and version.endswith("]"):
        return version[1:-1]

    version_range = versionexpr.replace(",", " ")
```

Anyway, a simple _regex_ will do the work, something non-greedy like `^\[.+?(,\s*strict)?(,\s*prerrelease)?\]$`, and group matching. I don't feel like the format is so important if `strict` and `prerrelease` are not valid in a version string (maybe we could change them to `loose` and `include_prerrelease` to be safer).

climblinne replied:
Updated to latest comments

climblinne commented:
@jgsogo: Thanks for your review. I will do the tests later, but before I need some feedback for the syntax as described in the above comments

danimtb commented:
Commas are only used for version range `>1.1,<2.1` but not for OR conditions `>1.1 || 0.8`, not for equivalent version like `2.8` and not for semver compatible `~=3.0`. So first I would like to clarify if the proposed `strict` and `prerelease` options apply to all the configurations above or not.

AFAIK a expression like this would not be possible 🚫 `>1.1,<2.0,>1.2 || 0.8`  , right?

jgsogo commented:
I'm trying things here (https://semver.npmjs.com/) and in fact, commas are never used (we remove them in our code). Anyway, we should keep the possibility of having one comma (it is documented), but remove it for Conan 2.0. 

About the parameters, they have these meanings (from the docs):

```
-l --loose
        Interpret versions and ranges loosely

-p --include-prerelease
        Always include prerelease versions in range matching
```

With the `--include-prerelease` you allow the prereleased version to be taken into account (if not, only the ones that appear explicitely in the version_range will be considered). The `loose` allows some kind of minor-errors on version strings (although the output will always be a correct one).

climblinne commented:
> Commas are only used for version range `>1.1,<2.1` but not for OR conditions `>1.1 || 0.8`, not for equivalent version like `2.8` and not for semver compatible `~=3.0`. So first I would like to clarify if the proposed `strict` and `prerelease` options apply to all the configurations above or not.

The commas are just a Conan implementation and removed anyway, before using the semver library. 

The ´strict´ (loose=False) parameter apply to the version number. So ´1.2´ would be not a valid version, where ´1.2.0´ would be fine. The ´loose´ variant accept more non standard semver version numbers.

The ´prerelease´ option (include_prerelease) gives us the ability, that we can use e.g. development versions (1.2.1-dev.3) instead of old released version (1.2.0). It works on the range side and change the sorting of version numbers.

Both parameter are valid for all configurations. Only ´~=3.0´ I think is not a valid option, should be ´~3.0´.
> 
> AFAIK a expression like this would not be possible 🚫 `>1.1,<2.0,>1.2 || 0.8` , right?

Should be the same ´>1.2 <2.0 || 0.8´ , which would be fine.

climblinne commented:
Here are the options to go:
1. Use `include_prerelease=True` and ´loose=False´ (also `include_prerelease=False` and `loose=True` are possible, but they are the default values anyway) as parameters separated by a comma from the rest of the range set: `[>1.0, <2.0, loose=False, include_prerelease=True]`
2. Use `strict` (loose=False) and `prerelease` (include_prerelease=True) as keywords. Seperated by comma: `[>1.0, <2.0, strict, prerelease]`
3. same as 1, separated by a semicolon: `[>1.0, <2.0; loose=False; include_prerelease=True]`
4. same as 2, separated by a semicolon: `[>1.0, <2.0; strict; prerelease]`

Which way to go? Anything else?

jgsogo commented:
Yes, I think those are all the options. I vote to keep closer to the implementation, so I will go for: `[>1.0 || 0.8, loose=True|False, include_prerelease=True|False]`. It needs to keep backward compatibility (the _Conan comma_) and I will let any order in the arguments.

jgsogo commented:
After talking with the team, we agree that the verbose option should be the best/safest way to go. Final syntax examples should include:

```
libname/<version>@conan/channel
libname/[<version_range_expr>]@conan/channel
libname/[<min>, <max>]@conan/channel   # <-- This should work for backwards compatibility
libname/[<version_range_expression>, loose=True, include_prerelease=False]@user/channel
libname/[<version_range_expression>, include_prerelease=False]@user/channel
libname/[<version_range_expression>, loose=False]@user/channel
libname/[<version_range_expression>, include_prerelease=False, loose=True]@user/channel
...
```

by default, `loose=True` and `include_prerelease=False` as it is the default behavior in the library too.

PD.- I'm adding dependency on #3964 as that one should be fixed first (we cannot remove commas)

climblinne commented:
@jgsogo  Could you give me a hint, how to write test for protected method '_parse_versionexpr'?

jgsogo commented:
Yes, of course.

As we are going to separate tests into _functional_ and _unittests_ soon, maybe it is a good time to start. So, create a file `conans/tests/unittests/test_client/test_graph/test_range_resolver.py` (will need `__init__.py` for each new folder to become a python module). Then, we can create the unittests:

⚠️ Disclaimer: _following code may not be valid Python_

```python
import unittest
from conans.client.graph.range_resolver import _parse_versionexpr

class ParseVerionExpr(unittest.TestCase):

    def test_backwards_compatibility(self):
        self.assertEqual(_parse_versionexpr("2.3, 3.2"), ("2.3 3.2", True, False))
        self.assertEqual(_parse_versionexpr("2.3, <=3.2"), ("2.3 <=3.2", True, False))

    def test_only_loose(self):
        self.assertEqual(_parse_versionexpr("2.3 ,3.2, loose=True"), ("2.3 3.2", True, False))
        self.assertEqual(_parse_versionexpr("2.3 3.2, loose=False"), ("2.3 3.2", False, False))
        self.assertEqual(_parse_versionexpr("2.3 3.2, loose  = False"), ("2.3 3.2", False, False))
        self.assertEqual(_parse_versionexpr("2.3 3.2,  loose  = True"), ("2.3 3.2", True, False))
        ...

    def test_only_prerelease(self):
        self.assertEqual(_parse_versionexpr("2.3, 3.2, include_prerelease=False"), ("2.3 3.2", True, False))
        self.assertEqual(_parse_versionexpr("2.3, 3.2, include_prerelease=True"), ("2.3 3.2", True, False))
        ...

    def test_both(self):
        self.assertEqual(_parse_versionexpr("2.3, 3.2, loose=False, include_prerelease=True"), ("2.3 3.2", False, True))
        self.assertEqual(_parse_versionexpr("2.3, 3.2, include_prerelease=True, loose=False"), ("2.3 3.2", False, True))
        ...

    def test_invalid(self):
        self. assertRaisesRegexp(<ExceptionType>, "expected msg", _parse_versionexpr, "2.3, 3.2, unexpected=True")
        self. assertRaisesRegexp(<ExceptionType>, "expected msg", _parse_versionexpr, "2.3, 3.2, loose=Other")
        self. assertRaisesRegexp(<ExceptionType>, "expected msg", _parse_versionexpr, "2.3, 3.2,")
        ...

```

Our code will be solid like a rock; and this tests, as they are running a very specific function, will run very fast (add as many tests as you want). And more important, as we know that any issue related to the version-range string is correctly handled in this function, we don't have to worry about invalid strings in other tests, each test class should be testing just one functionality. 💪

climblinne commented:
@jgsogo Do we need unit tests for `max_satisfying` function? This is already tested inside the library with lot's of tests.

climblinne commented:
There is the following error on the linux test server: `OSError: [Errno 28] No space left on device`. Could someone help?

jgsogo commented:
Thanks!

`max_satisfying` is a function from `semver` package, we have to assume that they test it before releasing a new version.

jgsogo commented on conans/test/unittests/test_client/test_graph/test_range_resolver.py:
```diff
@@ -0,0 +1,39 @@
+        self.assertEqual(_parse_versionexpr("2.3, 3.2, include_prerelease=True, loose=False", output),
+                         ("2.3 3.2", False, True))
+
+    def test_invalid(self):
```
I think that invalid inputs should raise a ConanException; otherwise, what should Conan do with them? What is the output of `max_satisfying` if provided with an invalid version range?

jgsogo replied:
I think that your comment should be in this _thread_
> I tested it directly and it delivers no results back and unluckily no exception. This should be put into a new issue to follow up later (when something changed in the lib?).

If `max_satisfying` returns nothing, the next line will raise a `KeyError` (`return candidates.get(result)`), so I think that it is worth handling the error or raise a meaninful exception (`ConanException` telling that the version range has an invalid format).

climblinne replied:
I added now an `ConanException`, when the range is invalid. Unluckily this is not working for all of your invalid ranges, because the semver library is accepted most of them, when `loose=True`. But this would be more a feature/bug in the semver library.

jgsogo replied:
Your exception is being raised if `semver.Range` raises. We cannot rely on `semver` satisfying our documented behavior; we have to check on our side and pass only what we consider a valid string to `semver`. We have to be at least as restrictive as `semver` is (and will be).

climblinne replied:
I am not able to check, if the range is valid on the Conan side, because I don't know exact the semver syntax. And as long we have to live with the "comma", we are not able to check this on our side. Please put this into another issue, so that this issue comes through. It's an really important feature for "CI".

jgsogo replied:
We have to check the string on the Conan side, it has nothing to do with `semver` library. In the future we may move to another _semver_ library and we will have to maintain backwards compatibility, so the valid/invalid checks should be implemented on our side (we are only delegating the `max_satisfying` calculation to the library, but we could do it ourselves too).

I'm trying to get a regex to validate this string. I have this one, but still fails with some use cases.

```
[a-zA-Z0-9_,<>=.\s~]{2,}?(\s*,\s*(loose|include_prerelease)\s*=\s*(True|False)){1,2}
```

jgsogo replied:
I don't know how to do it with only one regex... but take a look at this proposal (it is built on top of your PR, feel free to merge from my branch): https://github.com/jgsogo/conan/commit/9d5192368021ce4fd6561122813b4fe752b0d1d4

climblinne replied:
Look's good for me. I merged it into my branch.

jgsogo commented on conans/test/unittests/test_client/test_graph/test_range_resolver.py:
```diff
@@ -0,0 +1,39 @@
+        self.assertEqual(_parse_versionexpr("2.3, 3.2, loose=Other", output),
+                         ("2.3 3.2 loose=Other", True, False))
+        self.assertEqual(_parse_versionexpr("2.3, 3.2, ", output),
+                         ("2.3 3.2 ", True, False))
```
I will add some invalid tests for things like:

```
"2.3, 3.2, 1.2.3"
"2.3 3.2; loose=True, include_prerelease=True"
"2.3\t3.2" <-- will tabs be supported?
"loose=True, 2.3 3.3"
```

climblinne replied:
I tested it directly and it delivers no results back and unluckily no exception. This should be put into a new issue to follow up later (when something changed in the lib?).

jgsogo replied:
These are invalid inputs according to Conan docs, so they should behave exactly the same as a malformed version range. We cannot rely on `semver` doing the validation for us, they can change their behavior in future releases and they won't tell us.

climblinne replied:
Look at https://github.com/conan-io/docs/issues/937

climblinne commented:
@jgsogo Are you able to check, what happen to your linux build server (out of memory)?

climblinne commented:
@jgsogo Could you rerun the last test. This is not an issue from my side.

climblinne commented:
@lasote @jgsogo @memsharded Please check your reviews again. Thanks!

danimtb commented on conans/test/unittests/test_client/test_graph/test_range_resolver.py:
```diff
@@ -0,0 +1,50 @@
+        self.assertEqual(_parse_versionexpr("~1.2.3-beta.2", output), ("~1.2.3-beta.2", True, False))
+        self.assertEqual(_parse_versionexpr("^0.0", output), ("^0.0", True, False))
+        self.assertEqual(_parse_versionexpr("1.2.3 - 2.3.4", output), ("1.2.3 - 2.3.4", True, False))
+        self.assertEqual(_parse_versionexpr(">=1.2.3 <1.(2+1).0", output), (">=1.2.3 <1.(2+1).0", True, False))
```
I think this syntax is not valid in semver although  it is allowed in the regex. I think this comes from the **explanation** of the tilde ranges here: https://docs.npmjs.com/misc/semver#tilde-ranges-123-12-1

I have done some tests with this script and it is not working as expected:

```
from semver import SemVer, max_satisfying

filter = ">= 1.(1+1).0"

versions = [SemVer("1.0.0", loose=False), SemVer("1.2.0", loose=False)]
result = max_satisfying(versions, filter)
print("Result:", result)
```
The output is ``Result: None``. Whereas if I change the filter to `` ">= 1.2.0"`` it outputs the right result ``Result: 1.2.0``

climblinne replied:
Yes @danimtb, you are right. The syntax stands for `~1.2.3`. This seems to be bug in the semver library. Let's issue there a bug report.

climblinne replied:
What now, remove it? Leave round brackets inside regex?

jgsogo replied:
Raise a `ConanException` if it is not valid syntax (remove `(` and `)` from regex).

climblinne commented on conans/client/graph/range_resolver.py:
```diff
@@ -1,23 +1,79 @@
+            version_range.append(expr)
+
+        if match_param:
+            if match_param.group(1) == 'loose':
```
Done

climblinne commented on conans/client/graph/range_resolver.py:
```diff
@@ -1,23 +1,79 @@
+
+def _parse_versionexpr(versionexpr):
+    expression = [it.strip() for it in versionexpr.split(",")]
+    assert 1 <= len(expression) <= 4, "Invalid expression for version_range '{}'".format(versionexpr)
```
Done

climblinne commented on conans/client/graph/range_resolver.py:
```diff
@@ -1,23 +1,79 @@
+    loose = True
+    version_range = []
+
+    for i, expr in enumerate(expression):
```
As this part is from @jgsogo he should decide. I think it will be clearer this way, because we also need the "invalid" part.

jgsogo replied:
I would love to see a one-shot regex 🏆

climblinne commented:
> It does:
> 
> * it removes an unused `output` in code and in tests
> * applies some formatting: ordering of imports, pep8, line lengths, etc.
> * simplified naming in unittests, not necessary to call them "test". Adding missing ``**init**.py```
> 
> Please check and tell me if something wrong, should I revert something? Thanks!
I put back the 'output' in the code and added again the warning message for #3964. The message was lost on the way. I also added some tests for it.
